refactor(cloudformation): rename ServiceName parameter to WorkloadName

bvtujo opened the pull request:
Renames `ServiceName` and associated constructs to `WorkloadName`. This is not a breaking change, as it does not change SSM parameters. 

It does, however, change the names of some internal structs for better compatibility with the upcoming scheduled task work. 

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bvtujo committed:
chore(cloudformation): rename ServiceName to WorkloadName

bvtujo committed:
chore(cf): rename ServiceName to WorkloadName

bvtujo committed:
Make sure tests pass

efekarakus commented:
Out of curiosity, can we make sure to deploy a service before these changes and again after these changes to make sure that nothing breaks?

bvtujo commented:
Prior to this change:
```sh
➜  oaas git:(master) ✗ copilot-mainline -v
copilot version: v0.3.0-39-ga91d56f
➜  oaas git:(master) ✗ copilot-mainline svc deploy
Only found one service in workspace, defaulting to: fe
Only found one environment, defaulting to: test
Sending build context to Docker daemon  15.38MB
Step 1/7 : FROM golang:1.13
 ---> d6f3656320fe
Step 2/7 : WORKDIR /otter
 ---> Using cache
 ---> 4ea6aba31e4a
Step 3/7 : COPY . .
 ---> Using cache
 ---> 0e9f3d5dce77
Step 4/7 : ARG GOPROXY=direct
 ---> Using cache
 ---> 5f51a2b537f1
Step 5/7 : RUN go build otter.go
 ---> Using cache
 ---> 117b93317703
Step 6/7 : ENTRYPOINT go run otter.go
 ---> Using cache
 ---> ccb1c8e8be14
Step 7/7 : EXPOSE 8080
 ---> Using cache
 ---> 2a50a95f1318
Successfully built 2a50a95f1318
Successfully tagged 898620229039.dkr.ecr.us-east-2.amazonaws.com/otter/fe:120db9d
Login Succeeded
The push refers to repository [898620229039.dkr.ecr.us-east-2.amazonaws.com/otter/fe]
fa7921514296: Layer already exists
2fc9ce77423f: Layer already exists
f7222951894a: Layer already exists
ec7523d82c84: Layer already exists
73b60240f5be: Layer already exists
7279468fdfad: Layer already exists
e5df62d9b33a: Layer already exists
7a9460d53218: Layer already exists
b2765ac0333a: Layer already exists
0ced13fcf944: Layer already exists
120db9d: digest: sha256:ec53dac203e9b007ab6b30a4a4f548fa5cc3ed54efc40f0551e09dcd1a0842fe size: 2425


✔ Deployed fe, you can access it at http://otter-Publi-44FUTHMYIJJW-1496031724.us-east-2.elb.amazonaws.com.
```

Head commit:
```sh
➜  oaas git:(master) ✗ copi -v
copilot version: v0.3.0-42-g7a2f679
➜  oaas git:(master) ✗ copi svc deploy
Only found one service in workspace, defaulting to: fe
Only found one environment, defaulting to: test
Sending build context to Docker daemon  15.38MB
Step 1/7 : FROM golang:1.13
 ---> d6f3656320fe
Step 2/7 : WORKDIR /otter
 ---> Using cache
 ---> 4ea6aba31e4a
Step 3/7 : COPY . .
 ---> Using cache
 ---> 0e9f3d5dce77
Step 4/7 : ARG GOPROXY=direct
 ---> Using cache
 ---> 5f51a2b537f1
Step 5/7 : RUN go build otter.go
 ---> Using cache
 ---> 117b93317703
Step 6/7 : ENTRYPOINT go run otter.go
 ---> Using cache
 ---> ccb1c8e8be14
Step 7/7 : EXPOSE 8080
 ---> Using cache
 ---> 2a50a95f1318
Successfully built 2a50a95f1318
Successfully tagged 898620229039.dkr.ecr.us-east-2.amazonaws.com/otter/fe:120db9d
Login Succeeded
The push refers to repository [898620229039.dkr.ecr.us-east-2.amazonaws.com/otter/fe]
fa7921514296: Layer already exists
2fc9ce77423f: Layer already exists
f7222951894a: Layer already exists
ec7523d82c84: Layer already exists
73b60240f5be: Layer already exists
7279468fdfad: Layer already exists
e5df62d9b33a: Layer already exists
7a9460d53218: Layer already exists
b2765ac0333a: Layer already exists
0ced13fcf944: Layer already exists
120db9d: digest: sha256:ec53dac203e9b007ab6b30a4a4f548fa5cc3ed54efc40f0551e09dcd1a0842fe size: 2425


✔ Deployed fe, you can access it at http://otter-Publi-44FUTHMYIJJW-1496031724.us-east-2.elb.amazonaws.com.
```

Cloudformation template updates correctly, shows the Param name changed. 

### Cloudformation Stack Deployed by new version
Click to expand.
<details>

```yaml
# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
AWSTemplateFormatVersion: 2010-09-09
Description: CloudFormation template that represents a load balanced web service on Amazon ECS.
Parameters:
  AppName:
    Type: String
  EnvName:
    Type: String
  WorkloadName:
    Type: String
  ContainerImage:
    Type: String
  ContainerPort:
    Type: Number
  RulePath:
    Type: String
  TaskCPU:
    Type: String
  TaskMemory:
    Type: String
  TaskCount:
    Type: Number
  HTTPSEnabled:
    Type: String
    AllowedValues: [true, false]
  LogRetention:
    Type: Number
  AddonsTemplateURL:
    Description: 'URL of the addons nested stack template within the S3 bucket.'
    Type: String
    Default: ""
  HealthCheckPath:
    Type: String
  TargetContainer:
    Type: String
  TargetPort:
    Type: Number
Conditions:
  HTTPLoadBalancer:
    !Not
      - !Condition HTTPSLoadBalancer
  HTTPSLoadBalancer:
    !Equals [!Ref HTTPSEnabled, true]
  HasAddons: # If a bucket URL is specified, that means the template exists.
    !Not [!Equals [!Ref AddonsTemplateURL, ""]]
  HTTPRootPath: # If we're using path based routing and use the root path, we have some special logic
    !Equals [!Ref RulePath, "/"]
Resources:
  LogGroup:
    Type: AWS::Logs::LogGroup
    Properties:
      LogGroupName: !Join ['', [/copilot/, !Ref AppName, '-', !Ref EnvName, '-', !Ref WorkloadName]]
      RetentionInDays: !Ref LogRetention

  TaskDefinition:
    Type: AWS::ECS::TaskDefinition
    DependsOn: LogGroup
    Properties:
      Family: !Join ['', [!Ref AppName, '-', !Ref EnvName, '-', !Ref WorkloadName]]
      NetworkMode: awsvpc
      RequiresCompatibilities:
        - FARGATE
      Cpu: !Ref TaskCPU
      Memory: !Ref TaskMemory
      ExecutionRoleArn: !Ref ExecutionRole
      TaskRoleArn: !Ref TaskRole
      ContainerDefinitions:
        - Name: !Ref WorkloadName
          Image: !Ref ContainerImage
          PortMappings:
            - ContainerPort: !Ref ContainerPort
          # We pipe certain environment variables directly into the task definition.
          # This lets customers have access to, for example, their LB endpoint - which they'd
          # have no way of otherwise determining.
          Environment:
          - Name: COPILOT_APPLICATION_NAME
            Value: !Sub '${AppName}'
          - Name: COPILOT_SERVICE_DISCOVERY_ENDPOINT
            Value: !Sub '${AppName}.local'
          - Name: COPILOT_ENVIRONMENT_NAME
            Value: !Sub '${EnvName}'
          - Name: COPILOT_SERVICE_NAME
            Value: !Sub '${WorkloadName}'
          - Name: COPILOT_LB_DNS
            Value:
              Fn::ImportValue:
                !Sub "${AppName}-${EnvName}-PublicLoadBalancerDNS" 
          - Name: BUCKET123_NAME
            Value:
              Fn::GetAtt: [AddonsStack, Outputs.bucket123Name]
          LogConfiguration:
            LogDriver: awslogs
            Options:
              awslogs-region: !Ref AWS::Region
              awslogs-group: !Ref LogGroup
              awslogs-stream-prefix: copilot
        
        
  ExecutionRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Effect: Allow
            Principal:
              Service: ecs-tasks.amazonaws.com
            Action: 'sts:AssumeRole'
      Policies:
        - PolicyName: !Join ['', [!Ref AppName, '-', !Ref EnvName, '-', !Ref WorkloadName, SecretsPolicy]]
          PolicyDocument:
            Version: '2012-10-17'
            Statement:
              - Effect: 'Allow'
                Action:
                  - 'ssm:GetParameters'
                Resource:
                  - !Sub 'arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/*'
                Condition:
                  StringEquals:
                    'ssm:ResourceTag/copilot-application': !Sub '${AppName}'
                    'ssm:ResourceTag/copilot-environment': !Sub '${EnvName}'
              - Effect: 'Allow'
                Action:
                  - 'secretsmanager:GetSecretValue'
                Resource:
                  - !Sub 'arn:aws:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:*'
                Condition:
                  StringEquals:
                    'secretsmanager:ResourceTag/copilot-application': !Sub '${AppName}'
                    'secretsmanager:ResourceTag/copilot-environment': !Sub '${EnvName}'
              - Effect: 'Allow'
                Action:
                  - 'kms:Decrypt'
                Resource:
                  - !Sub 'arn:aws:kms:${AWS::Region}:${AWS::AccountId}:key/*'
      ManagedPolicyArns:
        - 'arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy'
  

  TaskRole:
    Type: AWS::IAM::Role
    Properties:
      ManagedPolicyArns:
      - Fn::GetAtt: [AddonsStack, Outputs.bucket123AccessPolicy]
      AssumeRolePolicyDocument:
        Statement:
          - Effect: Allow
            Principal:
              Service: ecs-tasks.amazonaws.com
            Action: 'sts:AssumeRole'
      Policies:
        - PolicyName: 'DenyIAMExceptTaggedRoles'
          PolicyDocument:
            Version: '2012-10-17'
            Statement:
              - Effect: 'Deny'
                Action: 'iam:*'
                Resource: '*'
              - Effect: 'Allow'
                Action: 'sts:AssumeRole'
                Resource:
                  - !Sub 'arn:aws:iam::${AWS::AccountId}:role/*'
                Condition:
                  StringEquals:
                    'iam:ResourceTag/copilot-application': !Sub '${AppName}'
                    'iam:ResourceTag/copilot-environment': !Sub '${EnvName}'
  

  DiscoveryService:
    Type: AWS::ServiceDiscovery::Service
    Properties:
      Description: Discovery Service for the Copilot services
      DnsConfig:
        RoutingPolicy: MULTIVALUE
        DnsRecords:
          - TTL: 10
            Type: A
          - TTL: 10
            Type: SRV
      HealthCheckCustomConfig:
        FailureThreshold: 1
      Name:  !Ref WorkloadName
      NamespaceId:
        Fn::ImportValue:
          !Sub '${AppName}-${EnvName}-ServiceDiscoveryNamespaceID'
  

  Service:
    Type: AWS::ECS::Service
    DependsOn: WaitUntilListenerRuleIsCreated
    Properties:
      Cluster:
        Fn::ImportValue:
          !Sub '${AppName}-${EnvName}-ClusterId'
      TaskDefinition: !Ref TaskDefinition
      DesiredCount: !Ref TaskCount
      PropagateTags: SERVICE
      LaunchType: FARGATE
      NetworkConfiguration:
        AwsvpcConfiguration:
          AssignPublicIp: ENABLED
          Subnets:
            - Fn::Select:
              - 0
              - Fn::Split:
                - ','
                - Fn::ImportValue: !Sub '${AppName}-${EnvName}-PublicSubnets'
            - Fn::Select:
              - 1
              - Fn::Split:
                - ','
                - Fn::ImportValue: !Sub '${AppName}-${EnvName}-PublicSubnets'
          SecurityGroups:
            - Fn::ImportValue: !Sub '${AppName}-${EnvName}-EnvironmentSecurityGroup'
      DeploymentConfiguration:
        MinimumHealthyPercent: 100
        MaximumPercent: 200
      # This may need to be adjusted if the container takes a while to start up
      HealthCheckGracePeriodSeconds: 60
      LoadBalancers:
        - ContainerName: !Ref TargetContainer
          ContainerPort: !Ref TargetPort
          TargetGroupArn: !Ref TargetGroup
      ServiceRegistries:
        - RegistryArn: !GetAtt DiscoveryService.Arn
          Port: !Ref ContainerPort

  TargetGroup:
    Type: AWS::ElasticLoadBalancingV2::TargetGroup
    Properties:
      #  Check if your service is healthy within 20 = 10*2 seconds, compared to 2.5 mins = 30*5 seconds.
      HealthCheckIntervalSeconds: 10 # Default is 30.
      HealthyThresholdCount: 2       # Default is 5.
      HealthCheckTimeoutSeconds: 5
      HealthCheckPath: !Ref HealthCheckPath
      Port: !Ref ContainerPort
      Protocol: HTTP
      TargetGroupAttributes:
        - Key: deregistration_delay.timeout_seconds
          Value: 60                  # Default is 300.
      TargetType: ip
      VpcId:
        Fn::ImportValue:
          !Sub "${AppName}-${EnvName}-VpcId"

  LoadBalancerDNSAlias:
    Type: AWS::Route53::RecordSetGroup
    Condition: HTTPSLoadBalancer
    Properties:
      HostedZoneId:
        Fn::ImportValue:
          !Sub "${AppName}-${EnvName}-HostedZone"
      Comment: !Sub "LoadBalancer alias for service ${WorkloadName}"
      RecordSets:
      - Name:
          !Join
            - '.'
            - - !Ref WorkloadName
              - Fn::ImportValue:
                  !Sub "${AppName}-${EnvName}-SubDomain"
              - ""
        Type: A
        AliasTarget:
          HostedZoneId:
            Fn::ImportValue:
              !Sub "${AppName}-${EnvName}-CanonicalHostedZoneID"
          DNSName:
            Fn::ImportValue:
              !Sub "${AppName}-${EnvName}-PublicLoadBalancerDNS"

  RulePriorityFunction:
    Type: AWS::Lambda::Function
    Properties:
      Code:
        ZipFile: |
          'use strict';const aws=require("aws-sdk"),priorityForRootRule="50000";let defaultResponseURL,report=function(a,b,c,d,e,f){return new Promise((g,h)=>{const i=require("https"),{URL:j}=require("url");var k=JSON.stringify({Status:c,Reason:f,PhysicalResourceId:d||b.logStreamName,StackId:a.StackId,RequestId:a.RequestId,LogicalResourceId:a.LogicalResourceId,Data:e});const l=new j(a.ResponseURL||defaultResponseURL),m={hostname:l.hostname,port:443,path:l.pathname+l.search,method:"PUT",headers:{"Content-Type":"","Content-Length":k.length}};i.request(m).on("error",h).on("response",a=>{a.resume(),400<=a.statusCode?h(new Error(`Error ${a.statusCode}: ${a.statusMessage}`)):g()}).end(k,"utf8")})};const calculateNextRulePriority=async function(a){var b,c=new aws.ELBv2,d=[];do{const e=await c.describeRules({ListenerArn:a,Marker:b}).promise();d=d.concat(e.Rules),b=e.NextMarker}while(b);let e=1;if(0<d.length){const a=d.map(a=>"default"===a.Priority||a.Priority===priorityForRootRule?0:parseInt(a.Priority)),b=Math.max(...a);e=b+1}return e};exports.nextAvailableRulePriorityHandler=async function(a,b){var c,d,e={};try{switch(a.RequestType){case"Create":d=await calculateNextRulePriority(a.ResourceProperties.ListenerArn),e.Priority=d,c=`alb-rule-priority-${a.LogicalResourceId}`;break;case"Update":case"Delete":c=a.PhysicalResourceId;break;default:throw new Error(`Unsupported request type ${a.RequestType}`);}await report(a,b,"SUCCESS",c,e)}catch(d){console.log(`Caught error ${d}.`),await report(a,b,"FAILED",c,null,d.message)}},exports.withDefaultResponseURL=function(a){defaultResponseURL=a};
      Handler: "index.nextAvailableRulePriorityHandler"
      Timeout: 600
      MemorySize: 512
      Role: !GetAtt 'CustomResourceRole.Arn'
      Runtime: nodejs10.x

  CustomResourceRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Version: 2012-10-17
        Statement:
          -
            Effect: Allow
            Principal:
              Service:
                - lambda.amazonaws.com
            Action:
              - sts:AssumeRole
      Path: /
      Policies:
        - PolicyName: "DNSandACMAccess"
          PolicyDocument:
            Version: '2012-10-17'
            Statement:
            - Effect: Allow
              Action:
                - elasticloadbalancing:DescribeRules
              Resource: "*"
      ManagedPolicyArns:
        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole

  HTTPSRulePriorityAction:
    Condition: HTTPSLoadBalancer
    Type: Custom::RulePriorityFunction
    Properties:
      ServiceToken: !GetAtt RulePriorityFunction.Arn
      ListenerArn:
        Fn::ImportValue:
          !Sub "${AppName}-${EnvName}-HTTPSListenerArn"

  HTTPSListenerRule:
    Type: AWS::ElasticLoadBalancingV2::ListenerRule
    Condition: HTTPSLoadBalancer
    Properties:
      Actions:
        - TargetGroupArn: !Ref TargetGroup
          Type: forward
      Conditions:
        - Field: 'host-header'
          HostHeaderConfig:
            Values:
              - Fn::Join:
                - '.'
                - - !Ref WorkloadName
                  - Fn::ImportValue:
                      !Sub "${AppName}-${EnvName}-SubDomain"
      ListenerArn:
        Fn::ImportValue:
          !Sub "${AppName}-${EnvName}-HTTPSListenerArn"
      Priority: !GetAtt HTTPSRulePriorityAction.Priority

  HTTPRulePriorityAction:
    Condition: HTTPLoadBalancer
    Type: Custom::RulePriorityFunction
    Properties:
      ServiceToken: !GetAtt RulePriorityFunction.Arn
      ListenerArn:
        Fn::ImportValue:
          !Sub "${AppName}-${EnvName}-HTTPListenerArn"

  HTTPListenerRule:
    Type: AWS::ElasticLoadBalancingV2::ListenerRule
    Condition: HTTPLoadBalancer
    Properties:
      Actions:
        - TargetGroupArn: !Ref TargetGroup
          Type: forward
      Conditions:
        - Field: 'path-pattern'
          PathPatternConfig:
            Values:
              !If
                - HTTPRootPath
                -
                  - "/*"
                -
                  - !Sub "/${RulePath}"
                  - !Sub "/${RulePath}/*"
      ListenerArn:
        Fn::ImportValue:
          !Sub "${AppName}-${EnvName}-HTTPListenerArn"
      Priority: 
        !If
          - HTTPRootPath
          - 50000 # This is the max rule priority. Since this rule evaluates true for everything, we make sure it is last
          - !GetAtt HTTPRulePriorityAction.Priority

  # Force a conditional dependency from the ECS service on the listener rules.
  # Our service depends on our HTTP/S listener to be set up before it can
  # be created. But, since our environment is either HTTPS or not, we
  # have a conditional dependency (we have to wait for the HTTPS listener
  # to be created or the HTTP listener to be created). In order to have a
  # conditional dependency, we use the WaitHandle resource as a way to force
  # a single dependency. The Ref in the WaitCondition implicitly creates a conditional
  # dependency - if the condition is satisfied (HTTPLoadBalancer) - the ref resolves
  # the HTTPWaitHandle, which depends on the HTTPListenerRule.

  HTTPSWaitHandle:
    Condition: HTTPSLoadBalancer
    DependsOn: HTTPSListenerRule
    Type: AWS::CloudFormation::WaitConditionHandle

  HTTPWaitHandle:
    Condition: HTTPLoadBalancer
    DependsOn: HTTPListenerRule
    Type: AWS::CloudFormation::WaitConditionHandle

  # We don't actually need to wait for the condition to
  # be completed, that's why we set a count of 0. The timeout
  # is a required field, but useless, so we set it to one.
  WaitUntilListenerRuleIsCreated:
    Type: AWS::CloudFormation::WaitCondition
    Properties:
      Handle: !If [HTTPLoadBalancer, !Ref HTTPWaitHandle, !Ref HTTPSWaitHandle]
      Timeout: "1"
      Count: 0

  AddonsStack:
    Type: AWS::CloudFormation::Stack
    Condition: HasAddons
    Properties:
      Parameters:
        App: !Ref AppName
        Env: !Ref EnvName
        Name: !Ref WorkloadName
      TemplateURL:
        !Ref AddonsTemplateURL
```

</details>

efekarakus commented on internal/pkg/deploy/cloudformation/stack/workload.go:
```diff
@@ -47,7 +47,7 @@ type templater interface {
 }
 
-type svc struct {
+type wkld struct {
```
What do you think about adding a comment here for what workload means? as I don't think it'll be as obvious to other readers as service.

```suggestion
// wkld represents a containerized workload running on Amazon ECS.
// A workload can be a long-running service, an ephemeral task, or a periodic task.
type wkld struct {
```

efekarakus commented on internal/pkg/deploy/cloudformation/stack/workload.go:
```diff
@@ -139,19 +139,19 @@ func (s *svc) templateConfiguration(tc templateConfigurer) (string, error) {
 
-func (s *svc) addonsOutputs() (*template.ServiceNestedStackOpts, error) {
-	stack, err := s.addons.Template()
+func (w *wkld) addonsOutputs() (*template.ServiceNestedStackOpts, error) {
```
Are we planning to rename this struct in a separate PR? that sounds good to me btw just double checking

bvtujo replied:
Yes, i plan to go package by package in the renaming effort.

efekarakus commented on internal/pkg/deploy/cloudformation/stack/workload.go:
```diff
@@ -139,19 +139,19 @@ func (s *svc) templateConfiguration(tc templateConfigurer) (string, error) {
 		var noAddonsErr *addon.ErrDirNotExist
 		if !errors.As(err, &noAddonsErr) {
-			return nil, fmt.Errorf("generate addons template for service %s: %w", s.name, err)
+			return nil, fmt.Errorf("generate addons template for service %s: %w", w.name, err)
```
should we rename this error to be:
```suggestion
			return nil, fmt.Errorf("generate addons template for %s: %w", w.name, err)
```

efekarakus commented on internal/pkg/manifest/backend_svc.go:
```diff
@@ -13,7 +13,7 @@ import (
 
 const (
-	backendSvcManifestPath = "services/backend/manifest.yml"
+	backendSvcManifestPath = "workloads/backend/manifest.yml"
```
I feel like it still makes sense to have `services` somewhere in the path what do you think of this?
```suggestion
	backendSvcManifestPath = "workloads/services/backend/manifest.yml"
```
And the common files between workloads can go to: `"workloads/common/param.json.tmpl"`

efekarakus reviewed (commented):
Looks great! Thank you so much for these changes, just some recommendations below:

bvtujo committed:
reword generate addons template error

Co-authored-by: Efe Karakus <efekarakus@gmail.com>

bvtujo committed:
Add comment to clarify "wkld"

Co-authored-by: Efe Karakus <efekarakus@gmail.com>

bvtujo committed:
update tests to reflect new wording

bvtujo committed:
move services to workloads/services dir

efekarakus reviewed (approved):
🚀

mergify[bot] committed:
Merge branch 'mainline' into cf/nameparam
